modules: mbedtls: make key exchange Kconfigs depend on, not select

tomi-font opened the pull request:
Turn the MBEDTLS_RSA_FULL selects into depends on. This is how the other MBEDTLS_KEY_EXCHANGE_* Kconfig options are defined.

This is done to avoid circular dependencies.

At the same time update uses of the affected MBEDTLS_KEY_EXCHANGE_* Kconfig options to enable/disable the dependencies which used to be automatically handled.

tejlmand reviewed (dismissed):
`depends on` is generally recommended over `select`.
See: https://docs.zephyrproject.org/latest/build/kconfig/tips.html#select-pitfalls

So approving.

nordicjm committed:
west.yml: MCUboot synchronization from upstream

Update Zephyr fork of MCUboot to revision:
  990b1fcb367e27056b282f183e819964fdbfe907

  - f76fba70 boot: bootutil: swap_scratch: Fix issue with bricking
    device
  - 7253f01c boot: bootutil: Refactor erase functionality to fix
    watchdog feeding
  - a98bff9f boot: zephyr: kconfig: Fix issues and re-order
  - 1b2d261d boot: zephyr: flash_map: Fix unused argument
  - 413eb384 boot: zephyr: flash_map: Fix missing include
  - 15b36f91 boot: zephyr: kconfig: enable dependencies of Mbed TLS
    Kconfig option
  - f6e8e88a boot: bootutil: Move erase function location
  - 5ef87c79 boot: zephyr: kconfig: Fix BOOT_SWAP_USING_MOVE
    description
  - bc18d7da boot: boot_serial: Fix issue with CBOR and setting
    image state

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

tomi-font committed:
modules: mbedtls: make key exchange Kconfigs depend on, not select

Turn the MBEDTLS_RSA_FULL selects into depends on.
This is how the other MBEDTLS_KEY_EXCHANGE_* Kconfig options are defined.

This is done to avoid circular dependencies.

At the same time update uses of the affected MBEDTLS_KEY_EXCHANGE_*
Kconfig options to enable/disable the dependencies which used to be
automatically handled.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

nordicjm commented on modules/mbedtls/Kconfig.mbedtls:
```diff
@@ -46,28 +46,33 @@ comment "Supported key exchange modes"
 
 config MBEDTLS_RSA_C
 	bool "RSA base support"
+	default y if UOSCORE || UEDHOC
```
not sure if these are required for `UOSCORE || UEDHOC` or not, if so then those symbols should select these or (imply them) since defaults will only apply when the project is initially configured, if someone has a project, uses menuconfig to enable mbedtls, then used menuconfig to enable `UOSCORE` or `UEDHOC` then these defaults will not be enabled

tomi-font replied:
Yeah this should likely be changed. There are many occurrences of this in the Mbed TLS Kconfig options. I just didn't want to change the behavior here, and because `MBEDTLS_KEY_EXCHANGE_RSA_ENABLED` previously selected those, I added the `default y` to them. But agreeing it should be done some other way.

valeriosetti replied:
I did a quick search in `./modules/lib/uoscore-uedhoc` for the uOSCORE and uEDHOC implementations and it seems to me that while the MD module is used a bit (not much, only 4 occurrences), the legacy RSA is not. As far as I can tell PSA API is used for crypto operations there, so enabling legacy RSA code in Mbed TLS might be a leftover from the old times. I know this is a preexisting problem, but it might be a good opportunity to remove that dependency. 
FWIW I also agree with @nordicjm comment above, but as for my point I admit this is preexisting so in case you don't want to address these problems in this PR I'm OK. I will take care of it later then ;)

tomi-font replied:
Yeah, some cleanup would be good, but I'll leave that as a follow-up exercise.

jukkar reviewed (dismissed):
+1 to networking changes

valeriosetti commented on modules/mbedtls/Kconfig.mbedtls:
```diff
@@ -104,25 +109,25 @@ config MBEDTLS_PSK_MAX_LEN
-	select MBEDTLS_MD
-	select MBEDTLS_RSA_FULL
+	depends on MBEDTLS_MD
+	depends on MBEDTLS_RSA_FULL
```
Although I agree that `depends on` is to be preferred over `select` for Kconfig, I think we should draw a line between what the user/other subsystems need to `imply`/`select` and what should be automatically inferred by this Kconfig (based for example on PSA Crypto API availability).
Here for example `MBEDTLS_KEY_EXCHANGE_RSA_ENABLED` will depend on `MBEDTLS_RSA_FULL` which means the full legacy implementation of RSA, but is it required in all the cases (ex: also when PSA API are available)? What if I need only PKCS1 v2.1 and I don't want to include v1.5 because it's weak and I want to save some Flash space?

tomi-font replied:
I guess it makes sense. All I did here was convert from `select` to `depends on`. Previously `MBEDTLS_KEY_EXCHANGE_RSA_ENABLED` would `select` both `MBEDTLS_PKCS1_V15` and `MBEDTLS_PKCS1_V21`.

valeriosetti commented on modules/mbedtls/Kconfig.mbedtls:
```diff
@@ -46,28 +46,33 @@ comment "Supported key exchange modes"
 config MBEDTLS_KEY_EXCHANGE_ALL_ENABLED
 	bool "All available ciphersuite modes"
+	select MBEDTLS_MD
+	select MBEDTLS_RSA_C
+	select MBEDTLS_PKCS1_V15
+	select MBEDTLS_PKCS1_V21
```
This is a bit strange to me: `MBEDTLS_KEY_EXCHANGE_ALL_ENABLED` means that you are going to enable all key exchanges, but then there is also legacy RSA support which might not be strictly required on all platforms. (See [other comment](https://github.com/zephyrproject-rtos/zephyr/pull/89200/files#r2073105170))

tomi-font replied:
This is anyway a bulk enabling of Kconfig options that must really just be used for testing. Or how would you change it?

valeriosetti commented on modules/mbedtls/Kconfig.mbedtls:
```diff
@@ -46,28 +46,33 @@ comment "Supported key exchange modes"
-	select MBEDTLS_PKCS1_V15
-	select MBEDTLS_PKCS1_V21
+	def_bool y
+	depends on MBEDTLS_RSA_C && MBEDTLS_PKCS1_V15 && MBEDTLS_PKCS1_V21
```
Based on [this comment](https://github.com/zephyrproject-rtos/zephyr/pull/89200/files#r2073105170) I don´t know if `MBEDTLS_RSA_FULL` is what we're looking for. It was OK when it was selecting all the other RSA related features, but now it makes less sense IMO. What about renaming `MBEDTLS_RSA_FULL` to `MBEDTLS_RSA_SOME` and change the `depends on` as follows:
```suggestion
	depends on MBEDTLS_PKCS1_V15 || MBEDTLS_PKCS1_V21
```
This is based on the fact that  `MBEDTLS_PKCS1_V15` and `MBEDTLS_PKCS1_V21` are already gated by `MBEDTLS_RSA_C`.

valeriosetti commented on modules/mbedtls/Kconfig.mbedtls:
```diff
@@ -257,15 +262,13 @@ config MBEDTLS_CIPHER_ALL_ENABLED
 config MBEDTLS_SOME_AEAD_CIPHER_ENABLED
-	bool
-	default y
+	def_bool y
```
Minor comment.
While the change is technically OK, the scope of the commit is different so it would have been better to place it in a separate commits IMO. The same holds also for `MBEDTLS_SOME_CIPHER_ENABLED` and `MBEDTLS_PSA_CRYPTO_CLIENT`.

valeriosetti reviewed (dismissed):
Although I agree that `depends on` is to be preferred over `select` for Kconfig, I think we should draw a line between what the user/other subsystems need to `imply`/`select` and what should be automatically inferred by `Kconfig.mbedtls`.
Having Kconfigs as `depends on` instead of `select`-ed allows for maximum configurability, but OTOH the user/subsystem willing to use some key exchange will have to manually select all the required algorithms and key types. This is OK to me, but I've seen people complaining about this kind of complexity during meetings, so it maybe a bit overkilling.

tomi-font commented:
@d3zd3z @ceolin please review

carlescufi commented:
@valeriosetti  please revisit this PR, I dismissed your NACK because it was stale.

tomi-font commented:
> @valeriosetti please revisit this PR, I dismissed your NACK because it was stale.

I know he is on vacation this week. I don't think he'll oppose in any way to merging this if @d3zd3z or @ceolin approves.

ceolin commented on modules/mbedtls/Kconfig.mbedtls:
```diff
@@ -46,28 +46,29 @@ comment "Supported key exchange modes"
 
 config MBEDTLS_PKCS1_V15
 	bool "RSA PKCS1 v1.5"
+	default y if UOSCORE || UEDHOC
```
I feel like this `UOSCORE` logic is an inverting dependencies. Shouldn't `UOSCORE` select `MBEDTLS_PKCS1_V15` ?

I won't block it since there are other symbols following the same logic, but it is looking inverted to me.

tomi-font replied:
Agreed. @valeriosetti was saying he might take care of changing this.

valeriosetti replied:
Hi, sorry but I was out of office last week. Yes, I started some work to fix this dependency. It's not ready yet, but I hope to finalize it soon.
